### DISCUSSION TITLE
DevExtreme DxForm not working in Nuxt3 production mode

### DISCUSSION THREAD

**Discussion by AloisSeckar** [nuxt/nuxt#16889] [Questions] [answered]
After struggling a bit to adopt [DevExtreme UI Components](https://js.devexpress.com/) into Nuxt3 project, I used solution proposed by **@stenet** in [earlier discussion](https://github.com/nuxt/framework/discussions/7159)

Creating plugin for DevExtreme components works fine, but only for Button (DxButton) and Menu (DxMenu). When I was developing further and introduced Form (DxForm), I ran into trouble. Everyting works in local `npm run dev` mode, but after `npm run build` the production code is failing upon encountering `<DxForm>` tag giving me following error:

    devextreme.987bc1bd.mjs:4 TypeError: f(...)[t] is not a function
        at t._createComponent (devextreme.987bc1bd.mjs:5:181913)
        at t._createComponent (devextreme.987bc1bd.mjs:313:53709)
        at t._createComponent (devextreme.987bc1bd.mjs:5:3153)
        at t._renderLayoutManager (devextreme.987bc1bd.mjs:313:53279)
        at t._renderLayout (devextreme.987bc1bd.mjs:313:50008)
        at t._initMarkup (devextreme.987bc1bd.mjs:313:47350)
        at t._initMarkup (devextreme.987bc1bd.mjs:5:3153)
        at t._renderComponent (devextreme.987bc1bd.mjs:5:179230)
        at t._updateDOMComponent (devextreme.987bc1bd.mjs:5:183078)
        at t.endUpdate (devextreme.987bc1bd.mjs:5:183257)

Unfortunately this vague message doesn't really tell me much about the actual cause. Maybe I am just inexperienced in debugging JavaScript and you'll see it.

My `plugins\dx.client.ts` file looks like this:

    import { DxButton } from "devextreme-vue/button";
    import { DxMenu, DxItem } from 'devextreme-vue/menu';
    import { DxForm, DxGroupItem, DxSimpleItem, DxLabel, DxButtonItem } from 'devextreme-vue/form';

    export default defineNuxtPlugin((nuxtApp) => {
        nuxtApp.vueApp
          .component("DxButton", DxButton)
          // menu
          .component("DxMenu", DxMenu)
          .component("DxItem", DxItem)
          // form
          .component("DxForm", DxForm)
          .component("DxFormGroup", DxGroupItem)
          .component("DxFormItem", DxSimpleItem)
          .component("DxFormLabel", DxLabel)
         .component("DxFormButton", DxButtonItem);
    });

As I already said, DxButton, DxMenu and DxItem elements work perfectly, but DxForm does not. At first I thought the problem is I am using aliases for DxFrom* sub-components, but it is not the case. I can easily use custom alias for DxButton and it still works, or keep original names for form components and the form still crashing.

I have node `16.17.0 LTS` using devextreme and devextreme-vue `22.1.4` and running the latest Nuxt3 `3.0.0-rc.9`

**Comment by AloisSeckar** [comment, marked as answer]
So in the end the answer (or at least a work-around) is the quite easy - disable tree-shaking for DevExtreme components in Vite configuration like so:

    vite: {
        plugins: [{
            name: "no-treeshake",
            transform(_, id) {
                if (id.includes("ui/")) {
                    return { moduleSideEffects: "no-treeshake" };
                }
            }
        }]
    }

The only thing I don't understand is, why is that happening? Is there some error in DevExtreme library? Or is it a Nuxt/Vite/Rollup issue?

**Comment by stenet** [comment]
@AloisSeckar yes, this is a known issue in DevExtreme in combination with Vite. DevExpress has no plans (or resources) to fix this in 2022. So the only workaround at the moment is disabling tree-shaking.

See [https://supportcenter.devexpress.com/ticket/details/t1054272/vue3-react-vite-rollup-devextreme-fails-in-production-because-some-modules-do-not-pass](https://supportcenter.devexpress.com/ticket/details/t1054272/vue3-react-vite-rollup-devextreme-fails-in-production-because-some-modules-do-not-pass)

**Comment by Caissan** [comment]
End of March 2023. Still not fixed.

**Reply by stenet** [reply]
@Caissan I use DevExtreme along with Nuxt 3 in production. If you have a specific problem, let me know, maybe I can point you in the right direction.

**Reply by Caissan** [reply]
Thanks, but the solution to disabling tree-shaking helped me.
